Refactor, UI update, Feature add

kbembedded opened the pull request:
This is going to end up being quite a lot of changes overall so I'm opening this as a draft right now for any comments. This is still a WIP (but mostly done, tested, and working at this point) as there are a few scattered comments as notes for myself later. These will all get cleaned up before this PR is moved out of draft.

A few things I want to note:
- I'm not trying to stomp all over your codebase, many of these changes are for clarity in reading and maintenance long term. I've tried to keep to the same code style, so please let me know if I missed any of the formatting or naming use.
- I'd happily discuss any choices I've made and if you feel they arn't appropriate would roll them back or update them as needed.
- This is another step in getting to a point where I'd be comfortable adding the functionality that I want to see in this codebase (as discussed in #11). I think once this is done and cleaned up, I'll start adding more views and configuration features.

kbembedded commented:
Looks like you did some linter setup, thanks for that. I noticed my formatting differences and changed some stuff before I saw you already did.

I havn't re-run this code through a linter, just manually resolved conflicts and re-tested. I'll be sure to do that before I mark this as ready.

Still a WIP (though I'm starting to add more back in now to do configurations of pokemon at runtime). Let me know if you've got any feedback on the commits so far.

EstebanFuentealba commented:
Hello @kbembedded, thank you for contributing to the project!
Yes, I had to configure the linter in order to submit a pull request to the official Flipper Zero catalog at https://github.com/flipperdevices/flipper-application-catalog/pull/7.

To pass the linter, you should use `ufbt format` using https://github.com/flipperdevices/flipperzero-ufbt.

EstebanFuentealba commented:
Could you provide some images or videos of what you did? I haven't received my Flipper Zero yet, so I haven't been able to test it.

🙏

kbembedded commented:
Sure! See photos below. The only visual changes I've made so far were in: 48d2a0916539f43d5d293c63c704b461ceac311f that affected the pokemon select screen to show pokedex number (rather than species hex value) as well as be able to see the full name of every pokemon. I'm not planning on adjusting the existing visuals any further.

My roadmap for configuration is going to be to probably add a list menu as the main screen. From there, select a pokemon, its level, and advanced options (its nickname, OT name, OT ID, moveset, etc., all of which will have somewhat sane defaults per pokemon and not the same across all).

My end goal of this is to be able to not just fill out my pokedex, but, be able to play through the game (I'm a few hours in to Blue at the moment) and fill out my pokedex by trading in reasonable pokemon that I wouldn't be able to get with just a copy of Blue and/or a single GB. So I want enough configuration to be able to select a sane level, moveset, etc., but then also be able to do absurd things (like teach a fire type water moves). As well as be able to trade a pokemon to the flipper and back to evolve it.

Photos of current visual changes:
![Screenshot-20230627-111126](https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/482ed63b-c973-46db-8206-a78f3621b590)
![Screenshot-20230627-111132](https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/98268666-62ce-4ba1-9ca8-528a931a4bbb)
![Screenshot-20230627-111157](https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/8e625d4b-e9ac-427c-a63e-553cab741ac5)

R4g3D commented:
@kbembedded
Would there by any chance be a way to make a throw away trade with the flipper (e.g. a Pidgey for a Pidgey), but on the Flipper side, scrape the data received from the GB for the OT nickname and OT ID, and save that data to the SD card for later use in trades?

Also, how's the refactoring going? I'm super keen on the functionality you hope to include in the app.

kbembedded commented:
> @kbembedded Would there by any chance be a way to make a throw away trade with the flipper (e.g. a Pidgey for a Pidgey), but on the Flipper side, scrape the data received from the GB for the OT nickname and OT ID, and save that data to the SD card for later use in trades?

That is a neat idea, but I'm not immediately sure of the feasibility of it. I'll keep that in the back of my head as I start moving forward with adding features. The complex part is when, where, and how to load that in in a way that makes sense with the general workflow. But I would agree that being able to save settings like that is probably a good idea. If I do add something like that, it probably won't be until a bit later in feature additions.

Poking #7 and #8 here just so those issues have a link to this PR so I can follow up later once I start rolling out features.


> Also, how's the refactoring going? I'm super keen on the functionality you hope to include in the app.

Thanks for asking! Seeing others with interest in this as well helps keep me motivated. Unfortunately, due to a number of life events, DEF CON planning, etc.,  I'm only able to work on this on and off and havn't had a large block of time to allocate to this in a couple of weeks. I should have some time this weekend so long as there no other fires in my life (literally and figuratively) that take priority.

R4g3D commented:
@kbembedded @EstebanFuentealba 
I've been working on my own version of this. I've managed to get it to set the Pokémon name as if it were a caught Pokémon without a nickname. Also managed to set the correct Pokémon type and allow for 4 moves to be selected. I've got a level selector too which sets the level and XP of the Pokémon.

I'm currently working on setting the stats to be appropriate for the level. I have attack, defence, speed and special working. But I can't seem to figure out Max HP, how did you come up with the default hex value of 2C01 in data? I don't see any relation to the resulting level 300 Pokémon, or to the HP formula on most Wiki pages.

Any help would be appreciated.

R4g3D commented:
It seems I spoke too soon. I just tested some different Pokémon at different levels. The level and XP are working, but the stats get glitches out.

So to amend my previous question, any insight into how to calculate the values based on level for the below would really help:
- hp
- hp_ev
- atk_ev
- def_ev
- spd_ev
- special_ev
- iv
- max_hp
- atk
- def
- spd
- special

From my understanding from my research, IV is a random value that can't be changed, so I accept setting the IV to its maximum (0xFFFF) to skip over the RNG of getting a good base IV. EV is the amount of experience in each stat based on battled Pokémon, it maxes out at 65535, so what I've done is set all the traded Pokémon's stat EVs to ((65535/100)*level). This way it assumes you've gained perfect EV in each stat per the chosen level. This is where things go wrong I believe. I've then used the formulas online to calculate the Pokémon's HP and other stats using the base value, the EV and the IV. This sort of works to get the values that should be displayed when viewing a Pokémon in game. The problem is, the hex values in the data_block don't match these decimal values I've calculated, so I don't know how to go from the decimal value to the correct hex value that should be transmitted to the GameBoy.

Hope this explains where I am at better, look forward to your response.

EstebanFuentealba commented:
Hello, I'm not entirely sure because I built this app using information from various sources. However, according to [Bulbapedia's Wiki](https://bulbapedia.bulbagarden.net/wiki/Pok%C3%A9mon_data_structure_(Generation_I)), HP and Max HP consist of 2 bytes.

In the initial version, Big Endian was used for **HP** (`0x1 0x2C`), and for **Max HP**, (`0x2c 0x1`) in Little Endian.

https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/blob/2f3a658d14d0004338221008d1a25ef106883000/pokemon_data.h#L14-L15

It seems that @kbembedded resolved it by using only Little Endian. Therefore, to modify HP, should we only consider 2C?

Honestly, I don't know much about the topic, but perhaps Kris can provide us with a better explanation on how to make modifications.

R4g3D commented:
That solved it! I was working in bid-endian and not little-endian.

Awesome!!!
I can now trade a Pokémon and it has the correct stats for the given level. Obviously this is still a little broken since it has perfect IV and optimal EV for its level. Might work on an option to randomise the IV and set the EV to zero to be more like an actual caught Pokémon.

Thanks for the pointer, nowhere in my research did endian come into play. Probably wasn't looking in the right places.

EstebanFuentealba commented:
@R4g3D 👏 Congratulations, I just tested it, and it works great! Perhaps in the future, we could create a list of all the moves for each Pokémon and fill them automatically. But it works wonderfully!

kbembedded commented:
Awesome! I'm excited to try this out. Sorry I fell behind on getting these features in here. 

> So to amend my previous question, any insight into how to calculate the values based on level for the below would really help:

Yeah, this is something I was hoping to tackle. It in theory should be possible, but, there are different scales of exp per level (from my initial research) so in order to have things line up cleanly I think you'd need some kind of identifier for each pokemon on how they gain things from their exp scale to line everything up.

> That solved it! I was working in big-endian and not little-endian.

Yes, it is annoying that there is an endian difference between the flipper and pokemon. When working with these you would need to translate back and forth from the flipper to z80.

> Perhaps in the future, we could create a list of all the moves for each Pokémon and fill them automatically.

FWIW my plan was just to set up the initial move list, e.g. the main data struct of each pokemon contains their values for level 1, including their default moves. It wouldn't be impossible to add a list of level/moves for each pokemon, but then you run in to "which moves should you clobber" when a pokemon would have learned more than 4 moves at its current level. I believe there is some kind of table in the actual game that would scale that. But also it would probably be easier to just let the user select what moves they want, and by the way, here are the default moves this pokemon would know at level 1.


I hope to play with this soon! I'll hold off on further work until @R4g3D's work gets merged in. Let me know, @R4g3D, if you want me to look at/make some contributions to your working codebase before readying this PR here. Otherwise I'll twiddle my thumbs a bit and finally put together a little PCB to easily connect a GameBoy Link cable to a flipper (been dragging my feet on that too).

kbembedded commented:
Actually, @R4g3D, @EstebanFuentealba, I can probably open this PR up from a draft if you both would prefer that. Would likely cause some conflicts with @R4g3D's codebase, so keep that in mind. I also have no issues with rebasing and cleaning these up from @R4g3D's HEAD.

I got no preference either way, let me know what you all would prefer and we can make it all work.

R4g3D commented:
Hey @kbembedded :)

It's mainly up to @EstebanFuentealba really, I'm happy to make a fresh pull request or whatever's necessary. My code is definitely not the neatest or most optimal, but it works which counts! XD

I'm on a bit of a roll now, I've just added a new screen to select from 6 choices of hidden stats (IV/EV) which affects the stats of the Pokémon. So on the one end you can now trade a Pokémon like it was a real catch and on the other end you can cheat yourself a perfect Pokémon.

I thought about adding functionality to restrict moves to only those that apply to the selected Pokémon or force the most recently learned 4 moves based on level, but as a hacker I like the option to do what I want. XD

I'm still developing though, about to test how I can trade a Pokémon without a nickname as currently I have set the nickname to the chosen Pokémon's name, but I realised after the fact that from previous experience, is a Bulbasaur with the nickname "Bulbasaur" evolves, you now have an Ivysaur called "Bulbasaur". It would be nice to add a screen to insert your own nickname, and default to no nickname if I get that working, but I don't know how to code that view, I've only been copy pasting and modifying the Pokémon selector screen from the original code. Which is another optimisation that could be made in the Pokémon level selector view, could just be a user input, or slider like I've seen in other apps. I tried looking at other source code to understand, but they're all in C and this is C++ so didn't want to open that can of worms just yet. Possible future development.

EstebanFuentealba commented:
Of course! You can contribute and make changes as you wish; Pull Requests are welcome.

kbembedded commented:
@R4g3D in that case, why don't you get it to a point where you're happy with your changes, open a PR, and then I can follow up later with rolling those in to the refactor here? I don't want to step on your toes but I agree that multiple views can get a bit unwieldy. The old adage of "its easier to optimize debugged code than to debug optimized code" I think applies here to get features functional, and then refine them later. If you have any other preference on moving forward, I'm all ears too :D

In terms of name, level, moveset, exp, etc., I was planning on a single list view with other sub menu options (kind of like if you go through "Settings" on the Flipper) rather than a separate view for each individual . So if you're okay with me cleaning that up at a later date, I'm fine with that and letting you keep your momentum.

> I'm still developing though, about to test how I can trade a Pokémon without a nickname as currently I have set the nickname to the chosen Pokémon's name, but I realised after the fact that from previous experience, is a Bulbasaur with the nickname "Bulbasaur" evolves, you now have an Ivysaur called "Bulbasaur"

I'm curious if setting the pokemon name to just the `TERM_` marker will work for gen I. According to: https://bulbapedia.bulbagarden.net/wiki/Nickname#In_the_core_series_games

>  In Generation I, a name consisting of only spaces can be used. In subsequent generations, submitting such a name is treated as exiting without entering a nickname; if such a name is submitted, a newly obtained Pokémon will use its species name, and a Pokémon at the Name Rater will keep its current name.

> Unnicknamed Pokémon use their species name in place of a nickname.

These make it sound like, at least for red/blue/yellow, if you just set `TERM_` it should fall back to species name. In other generations, all spaces with a `TERM_` is treated as not setting a nickname.

R4g3D commented:
So I've tried each of the below:
`{.str = {0x00, 0x00, 0x00 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}},`
`{.str = {TERM_, 0x00, 0x00 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}},`
`{.str = {TERM_, TERM_, TERM_, TERM_, TERM_, TERM_, TERM_, TERM_, TERM_, TERM_}},`

All of which result in a Pokemon with no name, it's just blank.
Not really sure what else to try other than all 0xFF.

kbembedded commented:
There are a handful of control characters in Gen I, you might try others of these? https://bulbapedia.bulbagarden.net/wiki/Character_encoding_(Generation_I)#Control_characters

Otherwise, I'm not sure. Might need to do some googling and if you can't find anything, fire up pokemon in an emulator that offers memory viewing and see what that struct looks like in memory. I think those games use hard memory locations for that struct, and if so, Bulbapedia would have them buried somewhere in their pages.

kbembedded commented:
Also, you might want to make sure that the pokemon you are testing with match your game's nickname and trainer ID. Foreign pokemon I think keep the same name even after they evolve, even if they had none? I could be wrong there. Everything I'm seeing also says something like, pokemon with no nickname are given the proper species name after an evolution or some other kind of event. So, it might be possible that the `TERM_` char is working, but, game logic won't update that till some other event similar to the game recalculating levels whenever there is EXP gain.

kbembedded commented:
I also found this: https://bulbapedia.bulbagarden.net/wiki/Save_data_structure_(Generation_I)#Pok.C3.A9mon

> A Pokémon's name is a nickname if it does not perfectly match the default name for a Pokémon (typically all uppercase) with any unused bytes of the entry's 11-byte capacity filled with 0x50 terminators. 

So, maybe the nickname just needs to be the pokemon name in all caps?

R4g3D commented:
I'll have to do some testing.

I have logic in the code already to make the nickname the Pokémon's name in all caps, but I had the name end with TERM_ and then filled empty space with 0x00.

I will modify the code to hardcode the level and xp so that the traded Pokémon will level up after 1 battle, then test a few of these solutions. 

I'll get to it tomorrow though since it's past midnight here in South Africa and I've been working on this all day.

Will post my findings here when I have them.

R4g3D commented:
So I've resolved the nickname issue, turns out the `trade_data_block` object in `pokemon_data.h` had:
`{.str = {F_, l_, i_, p_, p_, e_, r_, TERM_, 0x00, 0x00}},`
When it should have been:
`{.str = {F_, l_, i_, p_, p_, e_, r_, TERM_, TERM_, TERM_, TERM_}},`
So 11 characters instead of 10, where the last character is always a `TERM_` character.
Only the first 10 chars can store data, and when that data is the name of the Pokemon in all caps, it is treated as no nickname. Tested this by trading a level 6 Caterpie, and then evolving it, it now gets renames to Metapod.

However, I did notice a seperate issue now. I traded a level 6 Caterpie, and after the trade, when I check the Pokemon's details, everything looks perfect. It has the correct name, level, hp, moves, stats, everything. But, when I go battle a wild pokemon and it gains XP, it leveled up, to level 5. And after battling again, I checked the required XP to level up again and it was some strange large number. After battling a few more times (should have only been 2 more battles, ended up being like 4 or 5), it leveled up to level 7 and evolved into a Metapod with the correct name and data.
Not sure how this is happening. Everything looks right so not even sure where to start on this one.

Also something strange, I booted up my emulator and checked the memory. I'm using little-endian for all 2 byte values, but in the emulators memory, all the values are stored in big-endian. Not sure if that is an artifact of it being an emulator running on x86/x64 or what.

Oh! And also, this may relate to the level issue I'm having, but on my emulator, I have a level 50 Blastoise, level 50 Venusaur, level 50 Charizard, level 50 Alakazam, level 55 Dragonite and level 48 Aerodactyl. For the Blastoise, Venusaur and Charizard, the level value at offset 0x21 is correctly set to 0x32. But their level at offset 0x03 is 0x2E which is 46. For Alakazam, the level in offset 0x21 is also correct but the one at offset 0x03 is now 0x30 which is 48. For Dragonite and Aerodactyl, both level values are correct. No idea what is going on here.

R4g3D commented:
So just tested the level issue with a level 5 Bulbasaur, it didn't drop any levels, but it did have the same issue with the required XP to level up being a huge number, and only after battling a few more times than should have been needed did it level up to level 6.

Doing some testing now to see if I can fix it.

kbembedded commented:
Okay!

@R4g3D inspired me to finally get off my ass and slam out some code. I spent some time learning about some Flipper UI stuff in an attempt to reduce the binary size and runtime memory footprint while folding in the awesome changes from @R4g3D.

Because @R4g3D and I were doing work around the same git timeline, a bunch of the refactor work I did originally would no longer apply. It was unfortunately easier to revert @R4g3D's commits and then start folding those features back in to match the refactor and UI work.

This is still not done, but, its in a good enough state to get others using it. There are still many things from @R4g3D patch series that I need to re-apply here, but that will need to be a bit more manual.

Current features:
- Ability to select pokemon
- Ability to set level via text input
- Ability to set arbitrary nickname (still need to implement defaulting to to pokemon's name in all caps as well as the Nidoran M/F symbols)
- Set each move individually from a list
- EV/IV selection
- Set OT ID#
- Set OT Name
- Fixed experience calculation bug that @R4g3D had

I don't have a TODO list at this time, I just need to step away from the code for a day or so at this point before I burn myself out again lol.

I don't have a runtime memory use diff, but the binary size shrunk by about 10kbyte. With the application as large as it is right now, decreasing the size is going to be important if Gen II trading were to also be implemented.

I welcome any and all comments at this time. I might clean this branch up a bit and fold back in the rest of @R4g3D before moving this PR out of draft. Any further work to optimize or clean up would probably come at a later date.

R4g3D commented:
Awesome work! I'll give it a try tomorrow. Keen to see the flow on the Flipper. 

How did you solve the level/EXP bug?

kbembedded commented:
I think the math was slightly off? It looks like there may have been a misstep in simplifying it (or wherever you got it from was incorrect). Bear in mind, I only did basic testing by confirming the EXP and level of traded pokemon vs growth rate against the table here: https://bulbapedia.bulbagarden.net/wiki/Experience#Experience_at_each_level and have not actually done any battles with a traded pokemon so the bug _might_ still be there. But in checking exp every trade it lined up with what was expected for the current level.

```
    if(pokemon_table[trade->app->current_pokemon].xp_group == 0) {
        exp = 1.25 * level * level * level;
    } else if(pokemon_table[trade->app->current_pokemon].xp_group == 1) {
        exp = (1.2 * level * level * level) - (15 * level * level) + (100 * level) - 140;
    } else if(pokemon_table[trade->app->current_pokemon].xp_group == 2) {
        exp = level * level * level;
    } else if(pokemon_table[trade->app->current_pokemon].xp_group == 3) {
        exp = 0.8 * level * level * level;
    }
```

vs

```
    case GROWTH_FAST:
        // https://bulbapedia.bulbagarden.net/wiki/Experience#Fast
        experience = (4 * level * level * level) / 5;
        break;
    case GROWTH_MEDIUM_FAST:
        // https://bulbapedia.bulbagarden.net/wiki/Experience#Medium_Fast
        experience = (level * level * level);
        break;
    case GROWTH_MEDIUM_SLOW:
        // https://bulbapedia.bulbagarden.net/wiki/Experience#Medium_Slow
        experience =
            (((level * level * level) * 6 / 5) - (15 * level * level) + (100 * level) - 140);
        break;
    case GROWTH_SLOW:
        // https://bulbapedia.bulbagarden.net/wiki/Experience#Slow
        experience = (5 * level * level * level) / 4;
        break;
```

R4g3D commented:
Oh, yeah you for sure should test it with battles. Mine looked perfect when the traded Pokémon's stats were viewed, but after the first recalculation from doing a battle, it got messed up. 

I'll play around with your version tomorrow, see the new flow and test a couple trades with battles and let you know.

kbembedded commented:
Which pokemon did you use for testing? Might be specific to one growth rate unless you saw it for all 4 growth rates?

R4g3D commented:
I tried both a Weedle and a Bulbasaur. Didn't check if they had different growth rates though. But you need to use low level Pokémon, it's easier to see the change in EXP or a drop in their level after a battle. I think my level 6 weedle "levelled up" to level 5.

kbembedded commented:
Wild.

So, you are exactly right in that a traded pokemon from your commits _looks_ like it should be perfectly fine, but as soon as you get experience it goes down a level and then all of the other calculations look bonkers.

Using the commits I've made, I can make a pokemon look exactly the same, but after getting (nearly the same) exp it still appears to be normal. I took screenshots

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/5f7c6b14-0adf-493b-a81c-faf627af36d8" width="200"/>

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/b9079448-8400-4dbb-ba3b-d7de06bcc7fe" width="200"/>

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/d286d061-85c6-43dd-80bb-7c4ec058db7b" width="200"/>

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/ae727a81-cdb0-45a7-9c46-9d04b616dc83" width="200"/>

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/7516e285-1355-4e93-b8d4-04a8eb5a1773" width="200"/>

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/ef578dc9-6989-48e8-b0c0-078bf0e6cb32" width="200"/>

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/ce32ccd8-5616-4c14-a2cb-0ccf09b38557" width="200"/>

Note the weird needed experience. I think this might be related to the underflow bug that can happen in Gen I if you have a < level 2 pokemon?



My commits:

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/3d772c86-1a6f-4d67-9832-54b384da6d8f" width="200"/>

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/83a61670-1df0-4a42-ba6c-c78d2b9d30db" width="200"/>

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/7e5786aa-7d74-4773-9598-7fa53b3b378d" width="200"/>

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/a50d9265-6757-4beb-84ad-6821780542b2" width="200"/>

<img src="https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/assets/5025152/d72b78c3-4cc5-4ad7-a651-bf41cea5c450" width="200"/>


I'd guess there is a maths error somewhere in your implementation, but I'm not sure where. 

I also should note that while the actual math works out to the same between our implementations. We're both using unsigned integers to store the values, however you are multiplying in floating point numbers without any kind of casting. I forget how this kind of situation works as I rarely deal with it (either using int math only or specifically using floating point variable types), but I don't think C will cast anything behind the scenes. That doesn't explain why a traded pokemon still shows correct exp and level however.

In any case, mystery solved? lol

kbembedded commented:
This should be ready for final review/merge.

This PR moves the UI from a view for each step (which requires manual placement of elements, manual handling of input, etc.)  to using SceneManager (which allows for grouping views and instantiating modules more easily, as well as using built-in input handling and element placement).

This also adds the ability to specify the OT name and ID#.

It seems like it cleans up the exp bug that @R4g3D experienced. I'm not exactly sure why, but I can no longer reproduce the bug. The only thing I can think of is a difference of all integer maths vs floating point maths using integer variables would cause that exp issue. Can successfully work with a level 2 pokemon (level 1 does not exist in game due to underflow glitch that can occur).

I didn't update the README with photos, I'll deal with that another day, I just wanted to roll this out so I could get to a point where I can play through the game with the trade settings I wanted.

Bumps the changelog with versions, I pushed tags to my branch but I'm not sure if the PR pulls them in. Need to update the flipper application catalog once this gets merged in so that gets updated to make it easier to get these changes distributed.

Doing some playtesting now and I've not had any issues thus far.

R4g3D commented:
So I just gave it a test, when picking the moves, it's a bit of a pain to scroll through so many moves, and to have to do it 4 times. Not sure if there is a way to allow the left and right arrows to page down/up? Might make it a bit easier. Also, when you get to the trading screen, there is no way to click back now, so I had to force reset the Flipper. You removed the unicode male female characters for the Nidorans, however this is gonna cause an issue when they evolve, since it won't be the default name of the Pokemon, unless I am wrong and haven't checked the code and it's actually using the correct characters.

I haven't tried actually trading and battling yet, haven't had tome unfortunately. But otherwise, I think this version has more features than my own and is more streamlined. I like the added feature of having a default moveset for each Pokemon, nice touch. I think if you were to add the page up/down to the move list and fix the back button on the trade screen, this pull request should be accepted by @EstebanFuentealba.

Can't do much more now though, got to go to bed for work tomorrow. Will play around with it more when I get time.

kbembedded commented:
> So I just gave it a test, when picking the moves, it's a bit of a pain to scroll through so many moves, and to have to do it 4 times. Not sure if there is a way to allow the left and right arrows to page down/up? Might make it a bit easier.

I totally agree with this and don't really have a better solution at the moment other than re-implementing code that already exists in flipper. If you press and hold up/down it will scroll faster, but, it can still be time consuming. The submenu module doesn't seem to allow capturing input, so I think know if this is fixable using submenu without a flipper firmware mod. I definitely like navigating your original implementation better, but I was trying to slim down the code. I'm happy to try and rework if you and @EstebanFuentealba think adjusting this would be a need to have at this point in time.


> Also, when you get to the trading screen, there is no way to click back now, so I had to force reset the Flipper. 

That... might be a bug. Let me play with that some more. It certainly should be possible to go back.


> You removed the unicode male female characters for the Nidorans, however this is gonna cause an issue when they evolve, since it won't be the default name of the Pokemon, unless I am wrong and haven't checked the code and it's actually using the correct characters.

See code/commit comments. I was actually unable to build your code without commenting out the unicode escape sequence you added as I got the same error as here: https://github.com/EstebanFuentealba/Flipper-Zero-Game-Boy-Pokemon-Trading/pull/17#issuecomment-1710652016 but with the `'\u2642'` escape sequence. I was able to compile with the actual unicode symbol, however. I'm not sure why the build environments are different, but I'm literally unable to test with the escape sequence. If you have any insight as to why I can't build with escape sequence, I'd be happy to give that a go.

But also, Nidoran M/F are in all three Gen I games, I figured it was an okay compromise for the time to say "if you want it with the default name you either need to catch one, or trade it in and rename it" since you can set OT Name + ID# to match yours.

kbembedded commented:
Fixed trade not being able to go back. Same bug existed in OT Name which was also fixed.

In regards to navigating moves faster, after looking around a bit more there is no real good way to fix it at the moment. I think the only in-app fixes if press and hold is too slow that I can think of would be:
- Forego the submenu use and either implement a separate view with its own input handler
- Make another submenu to sort moves by letter. e.g. when Move 1 is selected, there is a submenu list of letters A-Z, select a letter, and submenu is redrawn with all of the moves staring with that letter. Back would go back to the alphabet, OK would select that move.

kbembedded commented:
Latest changes adds an index to the moves.

When you select a move to change you are presented with a list of:
```
No Move
Default Move
A
B
C
...
n
```
(where n is the first letter of the last move, assuming the move list is in alphabetical order which it is at the moment)

If you select `No Move` or `Default Move` it sets that move and dumps back to the main menu. If you select a letter from the index, you are then shown all of the moves starting with that letter. You can go back one and continue browsing, or select a move to have that set as the selected move and then get jumped back to the main menu.

kbembedded commented:
Another thing I just learned while playing is that the traded pokemon's type _DOES_ matter. I didn't think it did, and commented as such.

What actually seems to happen is when you view the pokemon's stats in-game, it shows their default types. However, the defined types when traded do end up affecting battle. e.g. the `trade_party` struct uses fire/ghost, and traded pokemon don't seem to be able to be affect by physical attacks. It was kind of weird having a Charmander that coudn't get hit with a tackle.

Notably, types are _not_ affected by the [Box trick](https://bulbapedia.bulbagarden.net/wiki/Box_trick). So, before this gets merged in I want to add back in proper default typing like @R4g3D originally had, as well as another menu to change them if desired.

I knew the original games had some quirks, but man, this one is actually really surprising.

kbembedded commented:
I ended up doing a "while I'm in there" and am cleaning up some other things I see that I poorly implemented. I should have a new push here in a few days.

@R4g3D If you have any insight on the unicode escape sequence issue, or if you ended up doing some kind of specific setup to support it, I'd love to know. I get it with both `ufbt` for just the repo and `fbt` if I have the repo in `applications_user` of the firmware tree. Since I'm re-touching everything I'd like to try and fix the Nidoran issues.

R4g3D commented:
Sorry I haven't responded. Been chaotic with work my side.
I see you managed to fix the Nidoran naming issue. I've compiled the new version quick and it's looking really good.
Still haven't had a chance to actually do any trades, and probably wont today unfortunately.
It seems that you have reached feature parity with my version, with added features and streamlined the code and UI.
If you're happy with the current state now and @EstebanFuentealba is happy with it, I'm good with this version over mine.
I definitely like the alphabet selection screen for the move selection. Makes it even faster than before.

kbembedded commented:
> I see you managed to fix the Nidoran naming issue.

Unfortunately it is still not functional. I have no clue how you made it work. I'm still trying to get it functionally correct by playing with wchar_t. Though If I have to use full wchar_t strings that's going to massively increase the binary size. I might just end up hacking a non-printable byte to be a stand in.

R4g3D commented:
So I just compiiled my version and somehow I am getting the compilation errors I was getting before.
Not sure how they went away and then came back, so I guess I didn't solve the Nidoran issue.

I've just compiled your latest version, not sure what you changed, but the option to change the level is now gone.
I also just traded a Nidoran, and on the GB side it seems to work, it's only on the Flipper that it's not visible.
Again, referncing my version of the app, the male/female symbol was never displayed, I think it just became more obvious when the option to change the name was available.
So from what I can see, if you just fix the level seclection issue, it should be good to go, or am I missing something?

kbembedded commented:
> I've just compiled your latest version, not sure what you changed, but the option to change the level is now gone.

lmao! How TF did I manage that. I'll take a look, thanks for the feedback >_<


> I also just traded a Nidoran, and on the GB side it seems to work, it's only on the Flipper that it's not visible.

Unfortunately that's how it's going to be for now. Once I touch the readme I'll make a note of that in there that the Nidoran M/F defaults to working but it doesn't display the symbol.


I also was doing some testing and I _think_ I have a memory leak. I may be missing a free() somewhere so I'm wanting to resovle that too.

kbembedded commented:
:tired_face:

I think its ready now. Fixed the missing level select (somehow removed a line of code that threw it on the screen). Updated photos and readme. Updated the changelog.

Give it a go and let me know if there is anything else outstanding at this point.

EstebanFuentealba commented:
Hello Kris, thank you very much for this great contribution!

I was testing it, and when I start the program with ./fbt launch APPSRC=pokemon, in the selection view, it doesn't change the Pokémon when I press LEFT or RIGHT.

However, if I close it and open it without ./fbt launch ..., it works fine.

[![ITest](https://img.youtube.com/vi/E1JrwvLaoCc/0.jpg)](https://www.youtube.com/watch?v=E1JrwvLaoCc)

kbembedded commented:
I ran in to this too. I _think_ its a main firmware bug. Basically, if you press directions and then press `OK` that pokemon gets set in the main menu and therefore selected to trade.

When I tried to debug this, I found the view_draw_callback to never actually get called except upon entering. It normally is called once a tick (I think). For some reason, when you do a `./fbt launch APPSRC=pokemon` from within the firmware, the draw callback gets registered correctly but never gets called on regular ticks. When you load it from the menu manually, its fine.

Note that if you use `ufbt` this does _NOT_ happen. Which leads me to believe there is something quirky going on with `fbt` / main firmware.

kbembedded commented:
I can actually see this behavior in a different demo app: https://github.com/jamisonderek/flipper-zero-tutorials/blob/main/ui/basic_scenes/basic_scenes.c

Its unfortunately not something observable without a dev board/debug session. But when `launch`ed, the `view_port_draw` and `view_dispatcher_draw_callback` never get hit. In `pokemon` this is ultimately what leads to the render callback for select pokemon.

If I back out and launch manually, I see these functions get hit over and over again on a tick cycle.

The same thing also happens to the trade screen. If you go to trade, it stays forever on the `Connect GB` screen and never updates because the draw callback never gets hit.

If I have time to dig in to this, I might open an Issue on the main firmware repo.

kbembedded commented:
Definitely a Flipper FW issue. I went back to 0.75.0 and it doesn't happen. Gonna try and track it down.

xMasterX commented:
> Hello Kris, thank you very much for this great contribution!
> 
> I was testing it, and when I start the program with `./fbt launch APPSRC=pokemon`, in the selection view, it doesn't change the Pokémon when I press `LEFT` or `RIGHT`.
> 
> However, if I close it and open it without `./fbt launch ...`, it works fine.
> 
> [![ITest](https://camo.githubusercontent.com/f85609ae05806a329f824171357e7a31d1c855313c5546723ad64b24d2c25081/68747470733a2f2f696d672e796f75747562652e636f6d2f76692f45314a7277764c616f43632f302e6a7067)](https://www.youtube.com/watch?v=E1JrwvLaoCc)

Correct, issue appears because of wrong tutorials and example apps
When you start app from Apps menu, it has loading animation, which stays in background and calls view_port_update
When you launching from favs, or via fbt (does same thing) - it does not call view_port_update,
because of that, app SHould call view_port_update by itself!  But, not with viewdispatcher
Since your app is based on viewdispatcher - please take a look into this example of workaround
https://github.com/xMasterX/all-the-plugins/commit/83e7b7219ccd515f1eead30f834efaf8273b0d9a

xMasterX commented:
> Definitely a Flipper FW issue. I went back to 0.75.0 and it doesn't happen. Gonna try and track it down.

Older fw had loading animation on every app launch event, check out my previous message for details about that

EstebanFuentealba commented:
Thank you, @xMasterX . I've added an update in input_callback

EstebanFuentealba commented on views/select_pokemon.c:
```diff
@@ -87,7 +87,13 @@ static bool select_pokemon_input_callback(InputEvent* event, void* context) {
+        {
+            model->curr_pokemon = selected_pokemon;
+        },
+        true);
```
`true` if you want to emit view update "draw"

kbembedded commented:
> Older fw had loading animation on every app launch event, check out my previous message for details about that

That makes a lot of sense. I noticed called `launch` no longer does that loading animation, but when you load the application manually it still does the loading animation, even briefly.

I also opened a bug report and one of the devs did respond saying it was incorrect use of view and yeah, using with_view_model seems like the answer.

Yes, I am abusing with_view_model, and you are welcome to rip it all out (or ask me to). When I did that work originally, it seemed completely unnecessary as locking wasn't being used so I saw there was no reason to need this whole synchronizing system when nothing was actually being protected and everything was still fully functional.


@EstebanFuentealba I'm guessing that change fixed it?

kbembedded commented:
I'm currently working on cleaning up my View model abuse. I'm trying to limit it to just that so as to not scope creep this PR. There are other cleanups I want to do long term (to make some reuse easier here and any other projects that might involved gameboy link protocol, and streamline the trade process a bit) but I think those are out of scope for the time being.

kbembedded commented:
Re-implemented, correctly, `with_view_model` in both the trade and select views.

The select view, as @EstebanFuentealba implemented in b7a8c92, was correctly triggering a re-draw. The changes I made were to stop abusing the model view (it was previously setting model to PokemonFap** and pointing that to the main struct) and then correctly use it to update information relevant to the screen display/select view itself. Only saving those changes back to the main struct when relevant.

This was not so straightforward for the trade view. Most of the functions called during the lifetime of this view are in an interrupt context. This means with_view_model cannot trigger a draw callback. Instead, a periodic timer was set up on 100 ms intervals to call a function that would a with_view_model call with no changes just to trigger a draw update.

EstebanFuentealba commented:
@kbembedded  Can I merge now, right?

kbembedded commented:
Oh, yeah! Sorry. I'm done fiddling around in this branch (I've started fiddling in other branches :D) Merge away!
